src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs

```diff
@@ -276,16 +291,14 @@ private void TlsHandshakeStopCore(BaseConnectionContext connection, long startTi
         _tlsHandshakeDuration.Record(duration.TotalSeconds, tags);
     }
 
-    private static void InitializeConnectionTags(ref TagList tags, BaseConnectionContext connection)
+    private static void InitializeConnectionTags(ref TagList tags, in ConnectionMetricsContext metricsContext)
     {
-        if (connection.LocalEndPoint is { } localEndpoint)
+        if (metricsContext.ConnectionContext.LocalEndPoint is { } localEndpoint)
         {
             // TODO: Improve getting string allocation for endpoint. Currently allocates.
             // Possible solution is to cache in the endpoint: https://github.com/dotnet/runtime/issues/84515
             // Alternatively, add cache to ConnectionContext.
```

Comment:
Side note: this issue is closed as wont-fix. We could fix this now or file an issue to cache it ourselves.

```suggestion
            // Cache to ConnectionContext.
```